### PR TITLE
#patch Add warehouse specialisation mechanic

### PR DESCRIPTION
**Pull request by lamnguyenkhoa**

### CODE REVIEW

File: src/buildings/warehouse/warehouse_building.gd [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+func _ready():
+	super()
+	data.type = EnumAutoload.BuildingType.STORAGE
+
+
+func apply_upgrades():
+	if EnumAutoload.UpgradeId.CONSTRUCTION_LOGIC_ADV_LOGISTIC in ResourceManager.current_upgrades:
+			get_node("Range").scale = Vector2(1.5, 1.5)
+	# Wait 2 frame to make sure all Area2D changes are setup correctly
+	await get_tree().physics_frame
```
Comment (ChewieCB) [inline src/buildings/warehouse/warehouse_building.gd:25]: Could we maybe replace this with a less general `await`? Something like `Object`'s `property_list_changed` signal? Depends on what exactly we're waiting for here, I'm assuming for the comment an Area2D's shape has been changed and we're waiting for the physics process to update with the new collision data?

File: src/buildings/warehouse/warehouse_building.gd [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	await get_tree().physics_frame
+	check_for_adjacency_multiplier(data.type)
+
+func set_specialisation(_type: EnumAutoload.ResourceType):
+	specialized_type = _type
+	print("Changed")
+	match specialized_type:
+		EnumAutoload.ResourceType.AIR:
+			sprite.self_modulate = Color(1, 0.8, 0.8)
```
Comment (ChewieCB) [inline src/buildings/warehouse/warehouse_building.gd]: We could remove this `match` statement if you store those Color values in an array and index it by `specialized_type` directly, as long as the array of Colors follows the indexing of the ResourceType enum.

File: src/buildings/warehouse/warehouse_building.gd [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+			resource_data.air = base_capacity
+	return resource_data
+
+
+func enable_improved_preview():
+	if EnumAutoload.UpgradeId.CONSTRUCTION_LOGIC_STOCK_ANALYSIS in ResourceManager.current_upgrades:
+		get_node("Range/RangeSprite").visible = true
+
+func remove_improved_preview():
+	get_node("Range/RangeSprite").visible = false
```
Comment (ChewieCB) [inline src/buildings/warehouse/warehouse_building.gd]: Grab this node in an @onready var to be called here, `get_node` calls within class methods always feels wrong to me - easier to miss mistakes in the string when it's this far down.

File: src/ui/context_menu/warehouse_spec_menu.gd [hunk 4]
```diff
@@ -0,0 +1,27 @@
+extends MarginContainer
+
+var selected_warehouse: WarehouseBuilding = null
+var main_panel: BuildingInfoPanel = null
+
+func _ready() -> void:
+	main_panel = get_parent().get_parent().get_parent()
+
+func _on_water_button_pressed() -> void:
+	SoundManager.play_button_click_sfx()
```
Comment (ChewieCB) [inline src/ui/context_menu/warehouse_spec_menu.gd]: We could add a generic `_on_resource_button_pressed(resource_type: EnumAutoload.ResourceType)` method that takes a resource type as an input and call it from each of these methods. Just to reduce code repetition.

**Review by ChewieCB — CHANGES_REQUESTED** [review]
LGTM, couple of minor code nitpicks I'd like for maintainability but overall looks clean.

File: src/buildings/warehouse/warehouse_building.gd (hunk 1 above)
Comment (lamnguyenkhoa) [reply src/buildings/warehouse/warehouse_building.gd:25]: Yeah, this is for the case buy the increased range upgrade for warehouses. Built warehouse won't be able to increase the it's range Area2D and check for collision fast enough in the same frame so I made it wait a bit. But the `property_list_changed ` seem promising, I will try it to see.

Comment (lamnguyenkhoa) [reply src/buildings/warehouse/warehouse_building.gd:25]: I cant find other solution so we have to keep these 2 await line tho.

**Comment by lamnguyenkhoa** [comment]
Hi I made changes, can you check it again

**Review by ChewieCB — APPROVED** [review]

**Pull request merged**